Revamp CircleCI builds.

petermattis opened the pull request:
Move almost all of circle.yml into build/circle-*.sh. These scripts can
be run locally with build/circle-local.sh performing nearly identical
work to what takes place on CircleCI.

This work isn't quite complete as I haven't done anything with
deployment or "make acceptance" yet, but it is worth a look and
comments.

petermattis committed:
Revamp CircleCI builds.

Move almost all of circle.yml into build/circle-*.sh. These scripts can
be run locally with build/circle-local.sh performing nearly identical
work to what takes place on CircleCI.

tamird commented:
@petermattis ping time to close this?

tschottdorf commented:
why close? I thought we'd get to it eventually. Especially now that we have actual acceptance tests.

petermattis commented:
What Tobias said. I'm going to get to this eventually, it's just fallen lower on my priority list.

petermattis commented:
This is ready for another look. The new acceptance tests are now run vs the old one that our existing circle.yml is still running. A cached build runs in ~8-9min vs the ~24min that builds are taking on the existing circle.yml and ~13min that they were taking before something broke.

Note that we no longer run `make test`, only `make testrace`.

tamird commented on build/circle-deps.sh:
```diff
@@ -0,0 +1,60 @@
+grep -v '^cmd' GLOCKFILE | glock sync -n
+
+# Pretend we're already bootstrapped, so that `make` doesn't try to get us
+# started which is impossible without a working Go env.
```
is this still relevant?

petermattis replied:
Relevant but inaccurate. We have a working Go env, but don't want to run bootstrapping since we're handling glock syncing manually in order to get the build-cache stuff to work. I'll adjust the comment.

tamird commented:
LGTM. Should `run/local-cluster.sh` be removed?

tamird commented on build/circle-test.sh:
```diff
@@ -0,0 +1,66 @@
+    cd $(dirname $0)/../acceptance
+    time ./acceptance.test -test.v -test.timeout -5m
+else
+    echo "skipping acceptance tests on $(uname)"
```
can this mention how to run the acceptance tests? `make acceptance`, yeah?

petermattis replied:
Sure. Done.

tamird commented on acceptance/localcluster/docker.go:
```diff
@@ -146,6 +146,12 @@ func maybePanic(err error) {
 // container.
 func (c *Container) Remove() error {
+	if os.Getenv("CIRCLE_ARTIFACTS") != "" {
+		// HACK: Removal of docker containers fails on circleci with the error:
```
do we know why this happens?

petermattis replied:
Nope. I thought it was related to docker version, but circleci upgraded to docker 1.6 recently and is still experiencing the error. It's not a big deal for us, though as they tear down docker at the end of each run causing all containers to be implicitly removed.

tschottdorf replied:
this is because of CircleCI's setup. I had asked their support and they said something about it being due to docker running pseudo-root.

petermattis commented:
`run/local-cluster.sh` is still used by `build/push-docker-deploy.sh`. Let's leave fixing that to a subsequent change.

petermattis committed:
Disable pushing the cockroachdb/cockroach-dev image.

We're not building it anymore, so we can't push it. I suppose we could
manually build it here, but that would add ~16min to each build. Should
we really be pushing new images for every successful master build?

petermattis commented on build/push-docker-deploy.sh:
```diff
@@ -10,12 +10,9 @@ run/local-cluster.sh start
+for version in {latest, ${VERSION}}; do
+  # Pushing to the registry just fails sometimes, so for the time
+  # being just make this a best-effort action.
+  docker push cockroachdb/cockroach:${version} || true
```
@tschottdorf Can you weigh in on this? Should I build the cockroach-dev image on the master branch? I'm not sure continuous deployment like this makes sense for a database product.

petermattis replied:
Perhaps we should look into nightly builds: https://circleci.com/docs/nightly-builds

tschottdorf replied:
no, we shouldn't build it. This is a good one for nightlies as you suggest (along with tougher acceptance tests). All you need to get nightly builds is something that schedules a nightly webhook. I'm sure there's a service providing this for almost free.

tschottdorf replied:
We may not even need it any more. I doubt it's used much.

petermattis replied:
I'm going to leave this as-is for now (pushing `cockroachdb/cockroach` on master builds) and file an issue to either move to nightly builds or remove.

tschottdorf replied:
(just to clarify: my comment about the image not being necessary is about the dev image. the other image is fine on master for now, but can move to nightlies when we have them)

petermattis committed:
Strip carriage returns from test output. Why are these showing up?

tschottdorf commented on build/circle-deps.sh:
```diff
@@ -0,0 +1,61 @@
+
+gopath0="${GOPATH%%:*}"
+cachedir="${gopath0}/pkg/cache"
+tag="20150719-133944"
```
anything to say about this?

petermattis replied:
The `cockroach/builder` image is versioned using this tag. If we add something new to the image, you run `build/builder.sh push` and then update the tag here. I'll add a comment explaining this.

tschottdorf commented on build/circle-deps.sh:
```diff
@@ -0,0 +1,61 @@
+if ! docker images | grep -q "${tag}"; then
+    # If there's a base image cached, load it. A click on CircleCI's "Clear
+    # Cache" will make sure we start with a clean slate.
+    rm -f "$(ls ${cachedir}/builder*.tar 2>/dev/null | grep -v ${tag})"
```
is there no equivalent invocation of `find` that does this cleanly?

petermattis replied:
This isn't clean?!?

tschottdorf replied:
something like
`find "${cachedir}" -type f -name 'builder*.tar' -a ! -name "*${tag}*" -delete`?

petermattis replied:
That's cleaner?

tamird replied:
also `-a` is an illegal optional in BSD find. I guess it would have to be:
```
find "${cachedir}" -type f -name 'builder*.tar' -not -name "*${tag}*" -delete
```
if that works in GNU find.

petermattis replied:
Or we can just leave the current code which has been tested and works on both circleci and mac.

tschottdorf replied:
leaving it is fine with me. We all have different OCDs and not all of them must be entertained.

tschottdorf commented on build/circle-deps.sh:
```diff
@@ -0,0 +1,61 @@
+    fi
+fi
+
+HOME= go get -d -u github.com/cockroachdb/build-cache
```
`HOME="" go get ...` is probably easier to read.

petermattis replied:
Yeah, it's effectively unsetting the home directory so that go doesn't pick up CircleCI's `.gitconfig`.

tschottdorf commented on build/circle-test.sh:
```diff
@@ -0,0 +1,69 @@
+find "${outdir}" -name 'test*.log' -type f -exec \
+     grep -B 5 -A 10 -E "^\-{0,3} *FAIL|${match}" {} ';' > "${outdir}/excerpt.txt"
+if [ -s "${outdir}/excerpt.txt" ]; then
+    echo "FAIL: excerpt.txt is not empty"
```
could print the first, say, 100 lines of it. Users which aren't logged in on CircleCI won't be able to get the excerpt, and for small fry it's helpful to see it right away.

petermattis replied:
I've changed this to output both the full path to `excerpt.txt` and the first 100 lines.

tschottdorf commented:
LGTM. Good to see this go live!

tschottdorf commented:
ran this locally and had this

> + docker pull cockroachdb/builder:20150719-133944
Pulling repository cockroachdb/builder
Get https://index.docker.io/v1/repositories/cockroachdb/builder/images: dial tcp: lookup index.docker.io on 10.0.2.3:53: read udp 10.0.2.3:53: i/o timeout

it still proceeded with `make check` so looks like we've probably ignored an error code...

tschottdorf commented on build/circle-test.sh:
```diff
@@ -0,0 +1,69 @@
+echo "make testrace"
+time ${builder} make GITHOOKS= testrace \
+     RACETIMEOUT=5m TESTFLAGS='-v --verbosity=1 --vmodule=monitor=2' | \
+    tr -d '\r' | tee "${outdir}/testrace.log" | \
```
I think the `tr` should stay. At least locally it messes up the output. We need to see what causes it to be different on CI.

tschottdorf commented:
can you print the output directory for the log files at the end of the run? I just ran circle-local.sh and it failed with a non-empty excerpt, but I have no idea where that file is.

petermattis commented:
@tschottdorf Looks like `circle-local.sh` was not setting `-e` which is what caused `circle-test.sh` to run even though `circle-deps.sh` failed. Easy enough to fix.

petermattis committed:
Document usage of the cockroachdb/builder tag for versioning.

Add "set -e" to circle-local.sh.

petermattis commented:
@tschottdorf fyi, when running locally, `outdir` is `TMPDIR`. I'm also outputting the full path to `excerpt.txt` when it is non empty.

bdarnell commented on acceptance/localcluster/docker.go:
```diff
@@ -146,6 +146,12 @@ func maybePanic(err error) {
 // Remove removes the container from docker. It is an error to remove a running
 // container.
 func (c *Container) Remove() error {
+	if os.Getenv("CIRCLE_ARTIFACTS") != "" {
```
`os.Getenv("CIRCLECI") == "true"` looks like the official way to test for this (https://circleci.com/docs/environment-variables)

petermattis replied:
Oh, thanks. Changed.

bdarnell commented on build/circle-test.sh:
```diff
@@ -0,0 +1,71 @@
+# 3. Run "make testrace".
+match='^panic|^[Gg]oroutine \d+|(read|write) by.*goroutine|DATA RACE'
+echo "make testrace"
+time ${builder} make GITHOOKS= testrace \
```
It makes me a little nervous that we're losing the exit status from `make testrace`, and instead relying on the `excerpt.txt` check below. I'd rather save the exit status from `make` so that we can fail if it was non-zero (after all the output-munging steps below).

petermattis replied:
We're running with `set -e`, I thought that would cause the script to fail if any individual command fails. Is that not the case with pipelines?

tamird replied:
Sadly it's not; that's why we have `test ${PIPESTATUS[0]} -eq 0` on L14.

You can get that behavioru with `set -o pipefail` though.

bdarnell replied:
With pipelines, only the last command's status counts unless you also `set -o pipefail`. Tamir did this for a bunch of our scripts recently. But we don't want to fail as soon as `make testrace` fails, because we still want the xml postprocessing to occur. We need to capture the exit status without failing the whole script, or split things up into separate circle.yml statements (this is how we currently run the xml step regardless of the status of `make`)

tschottdorf replied:
That was one of the useful aspects of running through `circle.yml`, it would always run all of the test statements separately even if some failed. Maybe the `xml` conversion and posting an error can go into a `trap`.

petermattis replied:
I would think that if any of our tests fail we want to just stop there. I thought that is the way `circle.yml` worked. I could have sworn that when `make check` failed nothing else was run.

Regardless, I'm changing this so that it remembers the `make test` and `make testrace` status and exits with 1 if either are non-zero.

bdarnell replied:
In general we probably want to stop early when something fails, but the xml step is useless unless it runs in failure cases. (the excerpt step is also useful)

tamird replied:
that's an interesting point - you can probably avoid doing the xml business unless there was a failure

tamird replied:
Eh, nevermind, that's incorrect.

petermattis replied:
Ok, here is my plan:

* make check - exit early if this fails
* make generate - exit early if this fails
* make test
  - make testrace - run only if make test succeed
* generate excerpt.txt and xml reports (always performed if make test is run)

bdarnell commented:
LGTM.

Do we even build in non-race mode any more? (we must when we build the deploy image, right?) If we do, then the extra minute or so to run both `make test` and `make testrace` seems worthwhile (one exposes bugs that occur when something is too slow, and the other exposes bugs that occur if something is too fast). If we manage to avoid the non-race build, though, I don't think it's worth the build time to run everything twice.

petermattis commented:
@bdarnell Yes, we build non-race mode for acceptance tests. I can re-add `make test`, just adds a minute or two to the run.

petermattis committed:
Use CIRCLECI env variable instead of CIRCLE_ARTIFACTS.

Re-add running of "make test".

petermattis commented:
This is ready for another look. I believe I've addressed all of the feedback. Please let me know if I missed anything. FYI, with "make test" added to the mix, test runs now take ~10-11min. 

While I've tested and retested this approach, I wouldn't be at all surprised if something breaks when it is merged.

bdarnell commented on build/circle-deps.sh:
```diff
@@ -0,0 +1,65 @@
+
+set -ex
+
+if [ "$1" = "docker" ]; then
```
Add a comment about how this script is run first with no arguments then re-runs itself inside the container with the arg `docker` (actually, why not make this block a second script?)

petermattis replied:
Done.

bdarnell commented on build/circle-deps.sh:
```diff
@@ -0,0 +1,65 @@
+# Pretend we're already bootstrapped, so that `make` doesn't go
+# through the bootstrap process which would glock sync and run
+# build/devbase/deps.sh (unnecessarily).
+touch .bootstrap
```
Why is this here instead of in the `docker` portion? We haven't bootstrapped here, only inside the container.

We're not doing everything that `deps.sh` does - the `npm install` isn't happening until the `go generate` step later on, so we're not caching the javascript dependencies. The divergence between circle-deps.sh and deps.sh is worrying; under the old scheme deps.sh was run in both places. Is there any way to re-unify these?

petermattis replied:
No reason for this to be here. I've moved it to the `docker` portion.

Yes, the `npm install` isn't happening anymore. The divergence from `deps.sh` is unfortunate. Even if we did the `npm install` step here I don't think it would cache the dependency (it depends on where the npm stuff gets installed). Ideally, we would get the `npm install` stuff moved into the `cockroachdb/builder` image. I'll add a TODO comment.

bdarnell commented on build/circle-test.sh:
```diff
@@ -0,0 +1,104 @@
+
+# 1. Run "make check" to verify coding guidelines. 
+echo "make check"
+time ${builder} make GITHOOKS= check | tee "${outdir}/check.log"
```
Add `export GITHOOKS=` at the toplevel with a comment about why.

petermattis replied:
Done. It's necessary to prevent bootstrapping. If the `GITHOOK` files don't exist, bootstrapping is preformed.

bdarnell commented on build/circle-deps.sh:
```diff
@@ -0,0 +1,65 @@
+    # Build everything needed by circle-test.sh.
+    time go test -race -v -i ./...
+    time go install -v ${cmds}
+    time make install
```
Does this `make` need a `GITHOOKS=`?

petermattis replied:
Perhaps. It doesn't hurt.

bdarnell commented:
LGTM
